docs/changes/shared-plugins-during-build.md

```diff
@@ -15,8 +14,68 @@ Affect scope: `Vite Plugin Authors`
 
 ## Motivation
 
-// TODO:
+Align dev and build plugin pipelines.
 
 ## Migration Guide
 
-// TODO:
+To be able to share plugins across environments, plugin state must be keyed by the current environment. A plugin of the following form will count the number of transformed modules across all environments.
+
+```js
+function CountTransformedModulesPlugin() {
+  let transformedModules
+  return {
+    name: 'count-transformed-modules',
+    buildStart() {
+      transformedModules = 0
+    },
+    transform(id) {
+      transformedModules++
+    },
+    buildEnd() {
+      console.log(transformedModules)
+    },
+  }
+}
+```
+
+If we instead want to count the number of transformed modules for each environment, we need to keep a map:
+
+```js
+function PerEnvironmentCountTransformedModulesPlugin() {
+  const state = new Map<Environment, { count: number }>()
+  return {
+    name: 'count-transformed-modules',
+    perEnvironmentBuildStartEnd: true,
+    buildStart() {
+      state.set(this.environment, { count: 0 })
+    }
+    transform(id) {
+      state.get(this.environment).count++
+    },
+    buildEnd() {
+      console.log(this.environment.name, state.get(this.environment).count)
+    }
+  }
+}
+```
+
+To simplify this pattern, internally in Vite, we use a `useEnvironmentState` helper:
```

Comment:
This is somewhat related to `useEnvironmentState` helper. It's not directly related to this PR, but putting it here for now. I had to create a different plugin instance per-environment when using native plugins with Vite + Rolldown.

The implementation was like:
https://github.com/rolldown/vite/blob/f633bd2ceeaf3d3f28d6c9940977343b12f89ac6/packages/vite/src/node/plugin.ts#L339-L380
https://github.com/rolldown/vite/blob/f633bd2ceeaf3d3f28d6c9940977343b12f89ac6/packages/vite/src/node/plugins/index.ts#L85-L97

I'm not sure if this would be needed commonly, but maybe something similar to this is needed when using some of the  rollup plugins in npm.

Reply:
We discussed with Anthony about a similar helper when we were trying to see how to refactor the reporter plugin: https://github.com/vitejs/vite/pull/17292/files

The `api.getBuiltinPlugin` trick is interesting. I also think we're going to need a better story to make rollup plugins per environment. When we had the functional API for plugins it was very easy, just wrap it as `() => rollupPlugin()` and you're done. Same with the object based `{ environmentPlugins: () => rollupPlugin() }`.
We ended up ditching the first one because of back compat, and the idea in general (the second was ok regarding compatibility) because we didn't want to push for the concept of two types of plugins:
- Shared plugins (environment aware)
- Per-environment plugin

The user now needs to know that a per-env plugin can't be in the main pipeline. And that a shared plugin (that actually shares data across env) needs to be always used as a regular plugin.
But maybe it is unavoidable, because with the current API, we are just ignoring that all these rollup plugins exists. When we tried to make alias per plugin, we didn't have a good way to do it internally either.

I wonder if we should extend to `applyToEnvironment(environment) => boolean | plugin`, and so we can have something similar to what you did in rolldown-vite with a helper

```js
{
  name: 'per-environment-alias',
  applyToEnvironment(environment) => aliasPlugin(environment.config.alias)
}
```
If an object is returned instead of a boolean, it replaces the current plugin with it for this environment.

Reply:
> The user now needs to know that a per-env plugin can't be in the main pipeline. And that a shared plugin (that actually shares data across env) needs to be always used as a regular plugin.

Yeah, the user might sometime need to understand it. But I think that doesn't happen much and the decision of making shared plugins by default was good.

> I wonder if we should extend to `applyToEnvironment(environment) => boolean | plugin`, and so we can have something similar to what you did in rolldown-vite with a helper

That would work too. I went with
```js
makeEnvironmentAwarePlugin('per-environment-alias', env => aliasPlugin(env))
```
because it felt more intuitive from a plugin user standpoint for me (the metal model is that this function makes the plugin env-aware).